Implemented codegen for CTL record fix changes

gnumonik opened the pull request:
This branch contains the codegen tools needed for the CTL record fix updates (WIP)

Commit:
Implemented codegen for CTL record fix changes

cstml committed:
ran fix-files

bladyjoker commented:
@gnumonik please fix the build and notify us when it's done so we can review this.

bladyjoker reviewed (changes requested):
Please add tests where you compare against Haskell BuiltinData reps. Make sure to test `newtypes` and `records`

Commit:
ported over newtype fixes from the dead end crunch branch

Commit:
Upgraded CTL to my fork to test compatibility

Commit:
switched to my CTL fork, fixed imports, newtypes still broken

Commit:
passes local 'make ci'

Commit:
updated the shouldRender tests to conform with new CTL

Commit:
formatting

Commit:
newtype fixes (untested)

Commit:
created a 'mkPlutusNewtype' function to translate newtypes which get their To/FromData instances from GeneralizedNewtypeDeriving in Haskell. Renamed: mkSumTypeIndexed -> mkPlutusDataType, extremelyUnsafeMkSumType -> unsafeMkPlutusDataType. Fixed the Ledger type generation for ledger types which use GND.

Commit:
turned off Werror. Revert this commit later

Commit:
Implemented tests! They don't really work but they work *just enough* to 'run' them by hand and manually format the results! And it works! The data representations match! Please help me drazen...

gnumonik commented:
@bladyjoker need your help fixing the IPC so I don't have to literally run all the tests by hand. 

there's a lingering JSON issue with newtype records that we need to fix. 

but the stupid manual tests pass and that means this PR is just about done

bladyjoker committed:
Synced with latest CTL PR 405 and fixed some tests

bladyjoker committed:
Merge branch 'ctl-integration' of github.com:mlabs-haskell/purescript-bridge into bladyjoker/help-with-ctl-integration

bladyjoker committed:
Fixed RoundTrip `spago run` by adding node2nix and spago2nix

bladyjoker committed:
make fix-files

bladyjoker commented on test/RoundTrip/app/src/Main.purs:
```diff
@@ -34,6 +36,6 @@ main = do
         Right testData -> do
           error ""
-          log $ stringify $ encodeJson testData
+          log $ show (toData testData)
```
FIXME: THe json roundtrip test hangs here

Commit:
Switched to GeneralizedNewtypeDeriving for the newtypes in RoundTrip/Types to ensure compatibility w/ generated code

Commit:
Attempted to fix tests w/ conduit. Didn't work

bladyjoker committed:
Revert "Attempted to fix tests w/ conduit. Didn't work"

This reverts commit 14dbdafcd0648038102108c801287853a7c98bfa.

bladyjoker committed:
Scaffolded the round trip test suite

bladyjoker committed:
Full scaffold is here, broken bc Int/Integer conflict Argonatu vs PlutusData

bladyjoker committed:
Removed all mention of integers and now we have failing Data tests (WIN)

bladyjoker committed:
We're almost there see the error below

  test/RoundTrip/Spec.hs:158:26:
  1) Round trip tests (Purescript <-> Haskell), Round trip tests (Purescript <-> Haskell) with ledger bridge, should produce PlutusData compatible representations
       Falsifiable (after 1 test):
         PdEither (Right (Just False))
       hs> Wanted ResSuccess got ResError ps> Wanted Foreign PlutusData got error: (inj @"fromBytesError" "PlutusData.from_bytes() raised Deserialization failed in PlutusDataEnum.PlutusDataEnum because: No variant matched")on input: (hexToByteArrayUnsafe "643837613966643837613966643837393966643837393830666666666666")

  To rerun use: --match "/Round trip tests (Purescript <-> Haskell)/Round trip tests (Purescript <-> Haskell) with ledger bridge/should produce PlutusData compatible representations/"

bladyjoker committed:
Simplified in search of the problem

bladyjoker committed:
Woot woot we got passing tests

    should produce PlutusData compatible representations [✔]
      +++ OK, passed 100 tests.

bladyjoker committed:
Removed debug output

Round trip tests (Purescript <-> Haskell)
  Round trip prerequisite tests
    `[test/RoundTrip/app] $ spago build` should work [✔]
    `[test/RoundTrip/app] $ spago build` should not warn of unused packages buildable [✔]
  with defaultBridge
    should have a Purescript process running [✔]
    should produce Aeson-compatible representations [✔]
      +++ OK, passed 100 tests.
  with ledger bridge
    should have a Purescript process running [✔]
    should produce PlutusData compatible representations [✔]
      +++ OK, passed 100 tests.

Finished in 2.9049 seconds
6 examples, 0 failures

bladyjoker commented:
@gnumonik 

```
Round trip tests (Purescript <-> Haskell)
  Round trip prerequisite tests
    `[test/RoundTrip/app] $ spago build` should work [✔]
    `[test/RoundTrip/app] $ spago build` should not warn of unused packages buildable [✔]
  with defaultBridge
    should have a Purescript process running [✔]
    should produce Aeson-compatible representations [✔]    
      +++ OK, passed 100 tests.
  with ledger bridge
    should have a Purescript process running [✔]
    should produce PlutusData compatible representations [✔]    
      +++ OK, passed 100 tests.

Finished in 2.9049 seconds
6 examples, 0 failures
```

Commit:
Added Show (via GenericShow) instance generation for all Plutus bridge types + updated tests

Commit:
forgot to add the formatted/linted files

bladyjoker committed:
Revert "forgot to add the formatted/linted files"

This reverts commit b7659d429fec34168f3bbe0ac8b03fe636ba9e4c.

bladyjoker committed:
Refactored `code-quality.nix` and bypassed test/expected dir when fixing

bladyjoker committed:
Added `cabal run test:tests` to CI (using nix develop)

bladyjoker committed:
Fixed the cabal test environment by creating `nix develop .#roundTripTest`

bladyjoker committed:
CI build fix perhaps???

bladyjoker committed:
Try another CI run...

bladyjoker committed:
Used `beforeAll` combinator instead of `runIO`

bladyjoker committed:
Removed a separate rt test for `defaultBridge`

bladyjoker committed:
Fix missing `generated` directory

bladyjoker committed:
Bundled all tests together

bladyjoker committed:
Removed multiple `spago` invocations (perhaps some race condition?)

bladyjoker commented:
CI build is still broken, keeping it that way until I find a solution
